lecture_notes/2017-01-24.md

```diff
@@ -0,0 +1,451 @@
+# Topic: Higher-Order Flow-Analysis
+## Speaker: Olin Shivers
+
+### Intro / Motivation
+
+"Higher-Order Flow Analysis" is a better name (conveys an insight) than
+"Higher-Order Control-Flow Analysis"
+
+
+Start with data-flow analysis. It's a great idea, big hammer.
+Looking for path-invariant facts. Fact ~= theorem.
+
+> There's a flowchart on the board. Olin says he used to write BASIC
+> programs by first writing a flowchart.
+
+Example problem: available expressions. At line N, what expressions in the program
+have already been computed? (Have their value in a register, variables in the
+expression haven't been mutated.)
+
+> Olin: Every time through the loop, I've saved myself an "add". That's an optimization.
+
+
+### Program Analysis by Analogy
+
+Rice's theorem is great. "Any interesting question about Turing machines is unsolvable."
+Forget the halting problem, Rice's theorem is way better.
+
+> M: wait, what you're saying is you can't build an algorithm. But since logic
+>    is stronger than programs, you can reason in a logic about the problem.
+
+Very hard to find truths about a program.
+Suppose the space of all truths is a complex polygon. You **oculd** use logic
+to describe this space exactly, with a very big formula. Another way to describe
+the space is by enclosing it with a circle.
+
+Polygon is exact and very complicated.
+Circle is an approximation, but sound and very simple (origin, radius).
```

Comment:
I'd add a little more here, and also to the quote "it's not like the court."

I.e. we can circumscribe the polygon with a circle that captures everything in the polygon ("all the truth") but also extra/false stuff (plus some lies), or we can inscribe a circle within the polygon and avoid all the false stuff ("nothing but the truth") but then we miss some of the truth.